GAE/src/org/akvo/flow/rest/ReportRestService.java

```diff
@@ -0,0 +1,217 @@
+
+/*
+ * @startuml
+ * UI -> ReportRestService: POST
+ * ReportRestService --> UI: Report with id (JSON object in QUEUED state)
+ * ReportRestService -> ReportServlet: START
+ * ReportServlet -> FlowService: POST with Report id
+ * FlowService -> ReportRestService: PUT Report in IN_PROGRESS state
+ * UI -> ReportRestService: GET
+ * ReportRestService --> UI: List of Reports by current user (JSON list)
+ * FlowService -> ReportRestService: PUT Report (FINISHED_SUCCESS state and filename)
+ * UI -> ReportRestService: GET
+ * ReportRestService --> UI: List of Reports by current user (JSON list)
+ * UI -> FlowService: GET filename
+ * FlowService --> UI: Report file
+ * @enduml
+ */
+
+@Controller
+@RequestMapping("/reports")
+public class ReportRestService {
+
+    private final String[] doNotCopy = {
+            "user",
+            "createdDateTime",
+            "lastUpdateDateTime"
+          };
+
+    private ReportDao reportDao = new ReportDao();
+    private UserDao userDao = new UserDao();
+
+    /**
+     * Create a new Report from posted payload.
+     *
+     * @param requestPayload
+     * @return
+     */
+    @RequestMapping(method = RequestMethod.POST, value = "")
+    @ResponseBody
+    public Map<String, Object> saveNewReport(@RequestBody
+    ReportPayload payLoad) {
+        final ReportDto reportDto = payLoad.getReport();
+        final Map<String, Object> response = new HashMap<String, Object>();
+        ReportDto dto = null;
+        RestStatusDto statusDto = new RestStatusDto();
+        statusDto.setStatus("failed");
+
+        // if the POST data contains a valid ReportDto, continue.
+        // Otherwise, server will respond with 400 Bad Request
+        if (reportDto != null) {
+            User user = userDao.findUserByEmail(reportDto.getUser());
+
+            if (user != null) {
+                Report r = new Report();
+
+                BeanUtils.copyProperties(reportDto, r, doNotCopy);
+                r.setUser(user.getKey().getId());
+                r.setState(Report.QUEUED);
+                // Save it, so we get an id assigned
+                r = reportDao.save(r);
+
+                //Queue it
+                Queue queue = QueueFactory.getDefaultQueue();
+                TaskOptions options = TaskOptions.Builder.withUrl("/app_worker/reportservlet")
+                        .param(TaskRequest.ACTION_PARAM, ReportTaskRequest.START_ACTION)
+                        .param(ReportTaskRequest.ID_PARAM, Long.toString(r.getKey().getId()));
+                queue.add(options); //overwrite any supplied state
+                dto = new ReportDto();
+                DtoMarshaller.copyToDto(r, dto);
+                statusDto.setStatus("ok");
+            }
+        }
+
+        response.put("meta", statusDto);
+        response.put("report", dto);
+        return response;
+    }
+
+    // find all reports belonging to the current user
+    //TODO: get an unfiltered list if superAdmin?
```

Comment:
I think even the superAdmin should only get their own list of files.

Reply:
ok